IntelliJ IDEA Gradle sync with Zulu 11.0.7 JDK on macOS throws newly created daemon process has a different context than expected

From this report: https://youtrack.jetbrains.com/issue/IDEA-238037

Create the simplest possible gradle project and add a gradle wrapper (I tried 6.3 and 6.2.2).

```
plugins {
    java
}
```

Open the project (I used "Open..." and selected the directory containing the gradle files).

When Intellij does its Gradle sync using the Zulu distribution of 11.0.7, the sync fails:

```
2020-04-17 07:29:20,907 [44554852]   INFO - s.plugins.gradle.GradleManager - Instructing gradle to use java from /Users/mbp/.sdkman/candidates/java/11.0.7-zulu 
2020-04-17 07:29:20,908 [44554853]   INFO - s.plugins.gradle.GradleManager - Instructing gradle to use java from /Users/mbp/.sdkman/candidates/java/11.0.7-zulu 
2020-04-17 07:29:20,918 [44554863]   INFO - xecution.GradleExecutionHelper - Passing command-line args to Gradle Tooling API: -Didea.sync.active=true -Didea.resolveSourceSetDependencies=true -Porg.gradle.kotlin.dsl.provider.cid=46801207603351 -Pkotlin.include.android.dependencies=true --init-script /private/var/folders/14/gq1jbz355v3f9zx3mc1j0bz00000gn/T/ijinit1.gradle 
2020-04-17 07:29:22,593 [44556538]   INFO - .project.GradleProjectResolver - Gradle project resolve error 
org.gradle.tooling.GradleConnectionException: Could not run phased build action using Gradle distribution 'https://services.gradle.org/distributions/gradle-6.3-bin.zip'.
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:55)
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:29)
	at org.gradle.tooling.internal.consumer.ResultHandlerAdapter.onFailure(ResultHandlerAdapter.java:43)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor.lambda$run$0(DefaultAsyncConsumerActionExecutor.java:57)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.gradle.launcher.daemon.client.DaemonConnectionException: The newly created daemon process has a different context than expected.
It won't be possible to reconnect to this daemon. Context mismatch: 
Java home is different.
Wanted: DefaultDaemonContext[uid=null,javaHome=/Users/mbp/.sdkman/candidates/java/11.0.7-zulu,daemonRegistryDir=/Users/mbp/.gradle/daemon,pid=31959,idleTimeout=null,priority=NORMAL,daemonOpts=-XX:MaxMetaspaceSize=256m,-XX:+HeapDumpOnOutOfMemoryError,--add-opens,java.base/java.util=ALL-UNNAMED,--add-opens,java.base/java.lang=ALL-UNNAMED,--add-opens,java.base/java.lang.invoke=ALL-UNNAMED,--add-opens,java.prefs/java.util.prefs=ALL-UNNAMED,-Xms256m,-Xmx512m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]
Actual: DefaultDaemonContext[uid=a2b6b868-028b-4718-bbed-fce8e71f89c3,javaHome=/Users/mbp/.sdkman/candidates/java/11.0.7-zulu/zulu-11.jdk/Contents/Home,daemonRegistryDir=/Users/mbp/.gradle/daemon,pid=33905,idleTimeout=10800000,priority=NORMAL,daemonOpts=-XX:MaxMetaspaceSize=256m,-XX:+HeapDumpOnOutOfMemoryError,--add-opens,java.base/java.util=ALL-UNNAMED,--add-opens,java.base/java.lang=ALL-UNNAMED,--add-opens,java.base/java.lang.invoke=ALL-UNNAMED,--add-opens,java.prefs/java.util.prefs=ALL-UNNAMED,-Xms256m,-Xmx512m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]

	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.connectToDaemonWithId(DefaultDaemonConnector.java:253)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.doStartDaemon(DefaultDaemonConnector.java:222)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.startDaemon(DefaultDaemonConnector.java:211)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.connect(DefaultDaemonConnector.java:133)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:145)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:99)
	at org.gradle.tooling.internal.provider.DaemonBuildActionExecuter.execute(DaemonBuildActionExecuter.java:51)
	at org.gradle.tooling.internal.provider.DaemonBuildActionExecuter.execute(DaemonBuildActionExecuter.java:35)
	at org.gradle.tooling.internal.provider.LoggingBridgingBuildActionExecuter.execute(LoggingBridgingBuildActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.LoggingBridgingBuildActionExecuter.execute(LoggingBridgingBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.ProviderConnection.run(ProviderConnection.java:203)
	at org.gradle.tooling.internal.provider.ProviderConnection.runPhasedAction(ProviderConnection.java:171)
	at org.gradle.tooling.internal.provider.DefaultConnection.run(DefaultConnection.java:249)
	at org.gradle.tooling.internal.consumer.connection.PhasedActionAwareConsumerConnection.run(PhasedActionAwareConsumerConnection.java:56)
	at org.gradle.tooling.internal.consumer.connection.ParameterValidatingConsumerConnection.run(ParameterValidatingConsumerConnection.java:62)
	at org.gradle.tooling.internal.consumer.DefaultPhasedBuildActionExecuter$1.run(DefaultPhasedBuildActionExecuter.java:78)
	at org.gradle.tooling.internal.consumer.DefaultPhasedBuildActionExecuter$1.run(DefaultPhasedBuildActionExecuter.java:70)
	at org.gradle.tooling.internal.consumer.connection.LazyConsumerActionExecutor.run(LazyConsumerActionExecutor.java:87)
	at org.gradle.tooling.internal.consumer.connection.CancellableConsumerActionExecutor.run(CancellableConsumerActionExecutor.java:45)
	at org.gradle.tooling.internal.consumer.connection.ProgressLoggingConsumerActionExecutor.run(ProgressLoggingConsumerActionExecutor.java:61)
	at org.gradle.tooling.internal.consumer.connection.RethrowingErrorsConsumerActionExecutor.run(RethrowingErrorsConsumerActionExecutor.java:38)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor.lambda$run$0(DefaultAsyncConsumerActionExecutor.java:55)
	... 6 more
2020-04-17 07:29:22,594 [44556539]   INFO - System.util.ExternalSystemUtil - External project [/Users/mbp/tmp/parent-proj/gradle-proj] resolution task executed in 1728 ms. 
2020-04-17 07:29:22,959 [44556904]   INFO - ge.ExternalProjectsDataStorage - Save external projects data in 4 ms 
```

However, this only happens with 11.0.7-zulu. Not 11.0.6-zulu, not 12, 13, or 14 from zulu, not 11.0.7 from Amazon (all installed via sdkman). I'm inclined to say it's a JDK bug, yet the gradle build works just fine on the command line with 11.0.7-zulu. The only issue I can find is that invoking gradle from IntelliJ fails, and only on macOS. The same trivial build works just fine with IntelliJ 2020.1 with 11.0.7-zulu on Linux.
```
IntelliJ IDEA 2020.1 (Ultimate Edition)
Build #IU-201.6668.121, built on April 8, 2020
Licensed to rust-base64 / Marshall Pierce
Subscription is active until December 26, 2020
For non-commercial open source development only.
Runtime version: 11.0.6+8-b765.25 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.15.4
GC: G1 Young Generation, G1 Old Generation
Memory: 1500M
Cores: 8
Registry: ide.max.recent.projects=100, debugger.click.disable.breakpoints=true
Non-Bundled Plugins: Key Promoter X, com.andrewbrookins.wrap_to_column, mobi.hsz.idea.gitignore, com.jetbrains.plugins.ini4idea, com.mnr.java.intellij.idea.plugin.base64helper, io.protostuff.protostuff-jetbrains-plugin, lermitage.intellij.extra.icons, name.kropp.intellij.makefile, org.antlr.jetbrains.st4plugin, org.intellij.plugins.hcl, org.jetbrains.plugins.rest, Pythonid, ru.adelf.idea.dotenv, ru.meanmail.plugin.requirements, org.jetbrains.kotlin, ru.yole.jitwatch-intellij, org.asciidoctor.intellij.asciidoc, org.toml.lang, some.awesome
```

I have also reproduced this with this Zulu JDK version.
![image](https://user-images.githubusercontent.com/3117044/79633987-81124580-8170-11ea-9ba7-a90c9ab4df56.png)

--- a/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java
+++ b/subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java
@@ -28,8 +28,6 @@ import org.gradle.internal.hash.HashCode;
 import org.gradle.internal.hash.Hashing;
 import org.gradle.internal.hash.HashingOutputStream;
 import org.gradle.internal.io.NullOutputStream;
-import org.gradle.internal.nativeintegration.filesystem.FileSystem;
-import org.gradle.internal.nativeintegration.services.NativeServices;
 import org.gradle.testing.internal.util.RetryUtil;
 import org.hamcrest.Matcher;
 
@@ -63,10 +61,10 @@ import java.util.jar.Manifest;
 
 import static java.nio.file.StandardCopyOption.COPY_ATTRIBUTES;
 import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotEquals;
-import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assume.assumeTrue;
 
@@ -495,15 +493,19 @@ public class TestFile extends File {
     }
 
     public TestFile createLink(File target) {
-        FileSystem fileSystem = NativeServices.getInstance().get(FileSystem.class);
-        if (fileSystem.isSymlink(this)) {
+        if (Files.isSymbolicLink(this.toPath())) {
             try {
                 Files.delete(toPath());
             } catch (IOException e) {
                 throw new UncheckedIOException(e);
             }
         }
-        fileSystem.createSymbolicLink(this, target);
+        try {
+            getParentFile().mkdirs();
+            Files.createSymbolicLink(this.toPath(), target.toPath());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
         clearCanonCaches();
         return this;
     }

--- a/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
+++ b/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
@@ -20,10 +20,12 @@ import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.integtests.fixtures.AvailableJavaHomes
 import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
 import org.gradle.internal.jvm.Jvm
+import org.gradle.test.fixtures.file.TestFile
 import org.gradle.util.GradleVersion
 import org.gradle.util.Requires
 import org.gradle.util.TestPrecondition
 import spock.lang.IgnoreIf
+import spock.lang.Issue
 
 
 class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
@@ -39,6 +41,29 @@ class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
         file("gradle.properties").writeProperties("org.gradle.java.home": symlinkedJdk.canonicalPath)
         expect:
         succeeds("help")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/16816")
+    @Requires(TestPrecondition.SYMLINKS)
+    def "can successful start after a running daemon's JDK has been removed"() {
+        // Zulu sets their Java distribution up like this
+        def installedJdk = Jvm.current().javaHome
+        def removedJdk = file("removed-jdk")
+        removedJdk.mkdir()
+        new TestFile(installedJdk).copyTo(removedJdk)
+
+        // start one JVM with the removed-jdk
+        file("gradle.properties").writeProperties("org.gradle.java.home": removedJdk.canonicalPath)
+        succeeds("help")
+
+        when:
+        // remove the other JDK
+        removedJdk.deleteDir()
+        // try to start with the other-jdk
+        file("gradle.properties").writeProperties("org.gradle.java.home": installedJdk.canonicalPath)
+        then:
+        succeeds("help")
+
     }
 
     @IgnoreIf({ GradleContextualExecuter.embedded }) // This test requires to start Gradle from scratch with the wrong Java version

--- a/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
+++ b/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
@@ -44,26 +44,24 @@ class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
     }
 
     @Issue("https://github.com/gradle/gradle/issues/16816")
-    @Requires(TestPrecondition.SYMLINKS)
     def "can successful start after a running daemon's JDK has been removed"() {
-        // Zulu sets their Java distribution up like this
         def installedJdk = Jvm.current().javaHome
-        def removedJdk = file("removed-jdk")
-        removedJdk.mkdir()
-        new TestFile(installedJdk).copyTo(removedJdk)
+        def jdkToRemove = file("removed-jdk")
+        jdkToRemove.mkdir()
+        new TestFile(installedJdk).copyTo(jdkToRemove)
 
-        // start one JVM with the removed-jdk
-        file("gradle.properties").writeProperties("org.gradle.java.home": removedJdk.canonicalPath)
+        // start one JVM with jdk to remove
+        file("gradle.properties").writeProperties("org.gradle.java.home": jdkToRemove.canonicalPath)
         succeeds("help")
 
         when:
-        // remove the other JDK
-        removedJdk.deleteDir()
-        // try to start with the other-jdk
-        file("gradle.properties").writeProperties("org.gradle.java.home": installedJdk.canonicalPath)
+        // remove the JDK
+        jdkToRemove.deleteDir()
+        // don't ask for the removed JDK now
+        file("gradle.properties").delete()
         then:
+        // try to start another build
         succeeds("help")
-
     }
 
     @IgnoreIf({ GradleContextualExecuter.embedded }) // This test requires to start Gradle from scratch with the wrong Java version

--- a/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
+++ b/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
@@ -43,6 +43,10 @@ class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
         succeeds("help")
     }
 
+    // This test deletes a JDK installation while the daemon is running.
+    // This is difficult to setup on Windows since you can't delete files
+    // that are in use.
+    @Requires(TestPrecondition.NOT_WINDOWS)
     @Issue("https://github.com/gradle/gradle/issues/16816")
     def "can successful start after a running daemon's JDK has been removed"() {
         def installedJdk = Jvm.current().javaHome

--- a/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
+++ b/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
@@ -51,14 +51,14 @@ class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
         new TestFile(installedJdk).copyTo(jdkToRemove)
 
         // start one JVM with jdk to remove
-        file("gradle.properties").writeProperties("org.gradle.java.home": jdkToRemove.canonicalPath)
+        executer.withJavaHome(jdkToRemove)
         succeeds("help")
 
         when:
         // remove the JDK
         jdkToRemove.deleteDir()
         // don't ask for the removed JDK now
-        file("gradle.properties").delete()
+        executer.withJavaHome(installedJdk)
         then:
         // try to start another build
         succeeds("help")

--- a/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
+++ b/subprojects/launcher/src/integTest/groovy/org/gradle/launcher/SupportedBuildJvmIntegrationTest.groovy
@@ -19,16 +19,30 @@ package org.gradle.launcher
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.integtests.fixtures.AvailableJavaHomes
 import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.internal.jvm.Jvm
 import org.gradle.util.GradleVersion
 import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
 import spock.lang.IgnoreIf
-import spock.lang.Unroll
 
-@Requires(adhoc = { AvailableJavaHomes.getJdks("1.6", "1.7") })
+
 class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
 
-    @Unroll
+    @Requires(TestPrecondition.SYMLINKS)
+    def "can start Gradle with a JDK that contains symlinks"() {
+        // Zulu sets their Java distribution up like this
+        def installedJdk = Jvm.current().javaHome
+        def symlinkedJdk = file("symlink-jdk")
+        installedJdk.listFiles().each {
+            symlinkedJdk.file(it.name).createLink(it)
+        }
+        file("gradle.properties").writeProperties("org.gradle.java.home": symlinkedJdk.canonicalPath)
+        expect:
+        succeeds("help")
+    }
+
     @IgnoreIf({ GradleContextualExecuter.embedded }) // This test requires to start Gradle from scratch with the wrong Java version
+    @Requires(adhoc = { AvailableJavaHomes.getJdks("1.6", "1.7") })
     def "provides reasonable failure message when attempting to run under java #jdk.javaVersion"() {
         given:
         executer.withJavaHome(jdk.javaHome)
@@ -41,7 +55,7 @@ class SupportedBuildJvmIntegrationTest extends AbstractIntegrationSpec {
         jdk << AvailableJavaHomes.getJdks("1.6", "1.7")
     }
 
-    @Unroll
+    @Requires(adhoc = { AvailableJavaHomes.getJdks("1.6", "1.7") })
     def "fails when build is configured to use Java #jdk.javaVersion"() {
         given:
         file("gradle.properties").writeProperties("org.gradle.java.home": jdk.javaHome.canonicalPath)

--- a/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java
+++ b/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java
@@ -59,12 +59,16 @@ public class DaemonCompatibilitySpec implements ExplainingSpec<DaemonContext> {
 
     private boolean javaHomeMatches(DaemonContext potentialContext) {
         try {
-            File potentialJava = Jvm.forHome(potentialContext.getJavaHome()).getJavaExecutable();
-            File desiredJava = Jvm.forHome(desiredContext.getJavaHome()).getJavaExecutable();
-            return Files.isSameFile(potentialJava.toPath(), desiredJava.toPath());
+            File potentialJavaHome = potentialContext.getJavaHome();
+            if (potentialJavaHome.exists()) {
+                File potentialJava = Jvm.forHome(potentialJavaHome).getJavaExecutable();
+                File desiredJava = Jvm.forHome(desiredContext.getJavaHome()).getJavaExecutable();
+                return Files.isSameFile(potentialJava.toPath(), desiredJava.toPath());
+            }
         } catch (IOException e) {
-            return false;
+            // ignore
         }
+        return false;
     }
 
     private boolean priorityMatches(DaemonContext context) {

--- a/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java
+++ b/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java
@@ -16,8 +16,11 @@
 package org.gradle.launcher.daemon.context;
 
 import org.gradle.api.internal.specs.ExplainingSpec;
+import org.gradle.internal.jvm.Jvm;
 
-import static org.gradle.internal.FileUtils.canonicalize;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
 
 public class DaemonCompatibilitySpec implements ExplainingSpec<DaemonContext> {
 
@@ -55,7 +58,13 @@ public class DaemonCompatibilitySpec implements ExplainingSpec<DaemonContext> {
     }
 
     private boolean javaHomeMatches(DaemonContext potentialContext) {
-        return canonicalize(potentialContext.getJavaHome()).equals(canonicalize(desiredContext.getJavaHome()));
+        try {
+            File potentialJava = Jvm.forHome(potentialContext.getJavaHome()).getJavaExecutable();
+            File desiredJava = Jvm.forHome(desiredContext.getJavaHome()).getJavaExecutable();
+            return Files.isSameFile(potentialJava.toPath(), desiredJava.toPath());
+        } catch (IOException e) {
+            return false;
+        }
     }
 
     private boolean priorityMatches(DaemonContext context) {

--- a/subprojects/launcher/src/test/groovy/org/gradle/launcher/daemon/context/DaemonCompatibilitySpecSpec.groovy
+++ b/subprojects/launcher/src/test/groovy/org/gradle/launcher/daemon/context/DaemonCompatibilitySpecSpec.groovy
@@ -15,7 +15,9 @@
  */
 package org.gradle.launcher.daemon.context
 
+
 import org.gradle.internal.nativeintegration.ProcessEnvironment
+import org.gradle.internal.os.OperatingSystem
 import org.gradle.launcher.daemon.configuration.DaemonParameters
 import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.gradle.util.ConfigureUtil
@@ -69,8 +71,14 @@ class DaemonCompatibilitySpecSpec extends Specification {
     }
 
     def "contexts with different java homes are incompatible"() {
-        client { javaHome = tmp.createDir("a") }
-        server { javaHome = tmp.createDir("b") }
+        client {
+            javaHome = tmp.createDir("client")
+            javaHome.file("bin", OperatingSystem.current().getExecutableName("java")).touch()
+        }
+        server {
+            javaHome = tmp.createDir("server")
+            javaHome.file("bin", OperatingSystem.current().getExecutableName("java")).touch()
+        }
 
         expect:
         !compatible
@@ -79,24 +87,25 @@ class DaemonCompatibilitySpecSpec extends Specification {
 
     @Requires(TestPrecondition.SYMLINKS)
     def "contexts with symlinked javaHome are compatible"() {
-        def dir = new File(tmp.testDirectory, "a")
-        dir.mkdirs()
-        def link = new File(tmp.testDirectory, "link")
-//        new TestFile(link).createLink(dir)
-        ["ln", "-s", dir, link].execute().waitFor()
+        // Make something that looks like a Java installation
+        def jdk = tmp.testDirectory.file("jdk").createDir()
+        jdk.file("bin/java").touch()
 
-        assert dir != link
-        assert link.exists()
-        assert dir.canonicalFile == link.canonicalFile
+        def linkToJdk = tmp.testDirectory.file("link")
+        linkToJdk.createLink(jdk)
 
-        client { javaHome = dir }
-        server { javaHome = link }
+        assert jdk != linkToJdk
+        assert linkToJdk.exists()
+        assert jdk.canonicalFile == linkToJdk.canonicalFile
+
+        client { javaHome = jdk }
+        server { javaHome = linkToJdk }
 
         expect:
         compatible
 
         cleanup:
-        assert link.delete()
+        assert linkToJdk.delete()
     }
 
     def "contexts with same daemon opts are compatible"() {

